### DISCUSSION TITLE
[Question] Any way to set decider first?

### DISCUSSION THREAD

**Discussion by acktsap** [spring-projects/spring-batch#4359] [Q&A] [answered]
In xml based job configuration, we can set decider first like this.

```xml
<job id="decisionJob1" >
    <decision id="decider1" decider="stepFlowDecider">
        <end on="EXIT" exit-code="COMPLETED"/>
        <next on="*" to="job1Step1" />
    </decision>

    <step id="job1Step1">
        <tasklet ref="createFileTasklet"/>
    </step>
</job>
```

But when using code based job configuration, no way to start with decider. `JobBuilder` only provides start methods taking `Step`, `Flow` only.

https://github.com/spring-projects/spring-batch/blob/972951a40af03c9c0553fb4c438f9aaeb1c9f1b5/spring-batch-core/src/main/java/org/springframework/batch/core/job/builder/JobBuilder.java#L53-L78

It there any way to set decider first on code based job configuration?

**Comment by baezzys** [comment]
If you use SimpleJobBuilder instead of JobBuilder, you can create a JobFlowBuilder that starts with a decider.

This is an example.

```
		JobExecutionDecider decider = new JobExecutionDecider() {
			private int count = 0;

			@Override
			public FlowExecutionStatus decide(JobExecution jobExecution, @Nullable StepExecution stepExecution) {
				count++;
				return count < 2 ? new FlowExecutionStatus("ONGOING") : FlowExecutionStatus.COMPLETED;
			}
		};
		JobFlowBuilder builder = new SimpleJobBuilder(new JobBuilder("flow", jobRepository)).start(decider);
```

I hope this example works for you.

**Reply by acktsap** [reply]
Thank's for your workaround. But i'm looking for high level api using `JobBuilder`. Not concrete class `JobFlowBuilder`. As long as we can set `JobExecutionDecider` first, it seems natural to be able to set by using `JobBuilder` class.

**Comment by fmbenhassine** [comment]
For questions and support requests, we started using GitHub discussions, see [Getting Help](https://github.com/spring-projects/spring-batch#getting-help). Therefore, I am moving this to a discussion. We will change it to an issue if we validate a bug or a missing feature.

**Comment by fmbenhassine** [comment, marked as answer]
I agree with @acktsap , if it is possible to start a job with a decision in XML, it should be possible to do it in Java as well using the `JobBuilder` API. The `JobFlowBuilder` already provides a [constructor to create a flow with a decider](https://github.com/spring-projects/spring-batch/blob/c4ad90b9b191b94cb4d21f90e5759bda9857e341/spring-batch-core/src/main/java/org/springframework/batch/core/job/builder/JobFlowBuilder.java#L42-L46), so this feature could be implemented   by adding the following two methods:

```diff

public class JobBuilder extends JobBuilderHelper<JobBuilder> {

++	public JobFlowBuilder start(JobExecutionDecider decider) {
++		return new FlowJobBuilder(this).start(decider);
++	}

}

public class FlowJobBuilder extends JobBuilderHelper<FlowJobBuilder> {

++	public JobFlowBuilder start(JobExecutionDecider decider) {
++		return new JobFlowBuilder(this, decider);
++	}

}
```

With that in place, we should be able to start a job flow with a decider in a consistent way as XML:

```java
@Bean
public Job job(JobRepository jobRepository, JobExecutionDecider decider) {
	return new JobBuilder("myJob", jobRepository)
				.start(decider)
				.build()// build the fow
			.build(); // build the job
}
```

I will change this discussion into a feature request and move it to the issue tracker. Contributions are welcome!
